### PR TITLE
[Metal] Refactor Shell unit-tests to allow runtime selection of client rendering API.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Adds support for Metal rasterizers in the unit-test harness. Tests that explicitly need metal can select the same by specifying `ClientRenderingAPI::kClientRenderingAPIMetal` to the `CreateShell` fixture method on Darwin platforms (`GTEST_SKIP` otherwise should work fine on all other platforms as no-op implementations are present for the linker).

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Fixes https://github.com/flutter/flutter/issues/41930

File: shell/common/tests/shell_test_surface.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (liyuqian) [inline shell/common/tests/shell_test_surface.h]: Is it possible to remove this class (and thus shell_test_surface.h/cc) and just merge the `CreateTestSurface/CreateRenderingSurface` into `ShellTestPlatformView`?

File: shell/common/tests/shell_test_surface_metal.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (liyuqian) [inline shell/common/tests/shell_test_surface_metal.h]: Could we just have a single `shell_test_surface_metal.h` (no `shell_test_surface_metal_impl.h`), and just add `shell_test_surface_metal.cc` or `shell_test_surface_metal_impl.mm` to build depending on the GN settings?

File: shell/common/tests/shell_test_surface.h (hunk 1 above)
Comment (chinmaygarde) [reply shell/common/tests/shell_test_surface.h]: This is the common interface used by `ShellTestSurfaceGL` and `ShellTestSurfaceMetal`. Without this common interface, the ShellTestPlatformView will have to own two unique pointers to both ShellTestSurfaceGL and ShellTestSurfacePlatform view and then conditionally check for the availability of either on each call. Now, the base can own just one member and call CreateRenderingSurface on it without null checks everywhere.

File: shell/common/tests/shell_test_surface_metal.h (hunk 2 above)
Comment (chinmaygarde) [reply shell/common/tests/shell_test_surface_metal.h]: We want the tests to be able to compile on both Darwin and non-Darwin platforms. Only that on non-Darwin platforms, the test will be skipped via GTEST_SKIP if Metal is not available and attempting to create a Metal surface will fail. shell_test_surface_metal.cc contains the no-op implementation that is used on non-Darwin platforms and shell_test_surface_metal_impl.mm contains the actual implementation on Darwin. Having just shell_test_surface_metal_impl.mm will cause linker errors on Linux and we cannot put ObjC code in a C++ in shell_test_surface_metal.cc.

File: shell/common/tests/shell_test_surface.h (hunk 1 above)
Comment (liyuqian) [reply shell/common/tests/shell_test_surface.h]: Can `ShellTestPlatformView` only own a unique pointer to `GPUSurfaceDelegate` (the parent class of both `ShellTestSurfaceMetalImpl` and `ShellTestSurfaceGL`), and directly implement `CreateRenderingSurface` based on that? We can discuss offline about this as I feel that Github comments don't have the bandwidth to resolve this efficiently.

File: shell/common/tests/shell_test_surface_metal.h (hunk 2 above)
Comment (liyuqian) [reply shell/common/tests/shell_test_surface_metal.h]: Discussed offline. The difference of those two approaches seems to be: `shell_test_surface_metal_impl.h` makes it easier to add metal only methods (such methods may be able to be just defined in the `.mm` file), while removing `shell_test_surface_metal_impl.h` makes it easier to add methods to `shell_test_surface.h`.

**Comment by chinmaygarde** [comment]
Rebased to ToT. PTAL. I don't remember if @liyuqian had a preference based on [his last comment](https://github.com/flutter/engine/pull/14023#discussion_r352793246). If this looks fine, can we land this. It is a rather large patch is hard to rebase to ToT.

cc @iskakaushik who was fiddling with the test vsync waiter than I just moved too.

**Comment by chinmaygarde** [comment]
> Ok to land this as is to avoid more merge conflict headaches

Too late already :( I'll rebase and remove the metal_impl.

**Pull request closed without merging**